streaming/src/main/scala/org/apache/gearpump/streaming/appmaster/ClockService.scala

```diff
@@ -16,54 +16,91 @@
  * limitations under the License.
  */
 
-package org.apache.gearpump.streaming.task
+package org.apache.gearpump.streaming.appmaster
 
 import java.util
 import java.util.Date
 import java.util.concurrent.TimeUnit
 
-import akka.actor.{Actor, Cancellable}
+import akka.actor.{Actor, Cancellable, Stash}
 import org.apache.gearpump.TimeStamp
+import org.apache.gearpump.streaming.AppMasterToExecutor.StartClock
+import org.apache.gearpump.streaming.appmaster.ClockService._
+import org.apache.gearpump.streaming.storage.AppDataStore
+import org.apache.gearpump.streaming.task._
 import org.apache.gearpump.streaming.{DAG, TaskGroup}
 import org.apache.gearpump.util.LogUtil
 import org.slf4j.Logger
 
+import scala.concurrent.Future
 import scala.concurrent.duration.FiniteDuration
-import scala.util.Try
 
 /**
  * The clockService will maintain a global view of message timestamp in the application
  */
-class ClockService(dag : DAG) extends Actor {
-import org.apache.gearpump.streaming.task.ClockService._
+class ClockService(dag : DAG, store: AppDataStore) extends Actor with Stash {
+  private val LOG: Logger = LogUtil.getLogger(getClass)
+
+  import context.dispatcher
 
   private val taskGroupClocks = new util.TreeSet[TaskGroupClock]()
   private val taskGroupLookup = new util.HashMap[TaskGroup, TaskGroupClock]()
-  private var scheduler : Cancellable = null
 
-  private val LOG: Logger = LogUtil.getLogger(getClass)
+  private var reportScheduler : Cancellable = null
+  private var snapshotScheduler : Cancellable = null
 
-  override def receive = clockService
+  override def receive = null
 
   override def preStart() : Unit = {
-    dag.tasks.foreach { taskIdWithDescription =>
-      val (taskGroupId, description) = taskIdWithDescription
-      val taskClocks = new Array[TimeStamp](description.parallelism).map(_ => Long.MaxValue)
-      val taskGroupClock = new TaskGroupClock(taskGroupId, Long.MaxValue, taskClocks)
-      taskGroupClocks.add(taskGroupClock)
-      taskGroupLookup.put(taskGroupId, taskGroupClock)
+    LOG.info("Initializing Clock service, get snapshotted StartClock ....")
+
+    store.get(START_CLOCK).asInstanceOf[Future[TimeStamp]].map { clock =>
+      val startClock = Option(clock).getOrElse(0L)
```

Comment:
For some self-driven TaskActor, like Split in Wordcount, they may send self a Message without Timestamp, so the clock will not update because the minimal number is always zero.